### PR TITLE
[TRTLLM-7967][feat] Adding Starcoder2 PyTorch Backend Support

### PR DESCRIPTION
**Pull request by yibinl-nvidia**
## Summary by CodeRabbit

* **New Features**
  * StarCoder2 model now available with support for 3B, 7B, and 15B parameter variants
  * Includes optimized sliding window attention mechanism for efficient inference
  * Full compatibility with CUDA graph acceleration for improved performance

* **Tests**
  * Added comprehensive validation and correctness testing for the new model

## Description

This PR implements PyTorch backend support for Starcoder2 3B, 7B, and 15B checkpoint, as well as the FP8 quantized checkpoint. Several tests are added to check network raw outputs or full e2e accuracy tests on GSM8K. 

Token level output comparision against HF implementation:
```
Prompt: 'def fibonacci(n):'                                            
                                                                       
TRT-LLM generated: '                                                                                                                          
    if n == 0:                                                                                                                                        
        return 0
    elif n == 1:                                                       
        return'
HF generated:      '
    if n == 0:
        return 0
    elif n == 1:
        return'
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tests/unittest/_torch/modeling/test_modeling_starcoder2.py [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+
+        starcoder2_config = Starcoder2Config.from_dict(config_dict)
+
+        dtype = starcoder2_config.torch_dtype
+        device = torch.device("cuda")
+
+        with torch.device(device), default_dtype(dtype):
+            model_config = ModelConfig(pretrained_config=starcoder2_config)
+            starcoder2 = Starcoder2ForCausalLM(model_config).to(device)
+
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/modeling/test_modeling_starcoder2.py]: There is no error, and torch.device(device) is the newer, more comprehensive, and recommended approach introduced in PyTorch 2.0.

File: tensorrt_llm/_torch/models/modeling_starcoder2.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+            preload_weight_modules = getattr(self, "preload_weight_modules", None)
+            _load_weights_impl_v2(
+                self,
+                weights,
+                weight_mapper,
+                skip_modules,
+                params_map=params_map,
+                preload_weight_modules=preload_weight_modules,
+            )
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/models/modeling_starcoder2.py]: fixed in the latest version.

File: tensorrt_llm/_torch/models/modeling_starcoder2.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+from tensorrt_llm._torch.modules.attention import Attention
+from tensorrt_llm._torch.modules.decoder_layer import DecoderLayer
+from tensorrt_llm._torch.modules.embedding import Embedding
+from tensorrt_llm._torch.modules.linear import TensorParallelMode
+from tensorrt_llm._torch.modules.mlp import MLP
+from tensorrt_llm._torch.speculative import SpecMetadata
+from tensorrt_llm.functional import PositionEmbeddingType
+
+
+class Starcoder2LayerNorm(nn.LayerNorm):
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_starcoder2.py]: TRTLLM has LayerNorm itself, can we use it? see tensorrt_llm/_torch/modules/layer_norm.py

File: tensorrt_llm/_torch/models/modeling_starcoder2.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        """
+        Load weights with custom mapping for StarCoder2.
+
+        StarCoder2 uses GPT-2 style MLP naming (c_fc, c_proj)
+        while our MLP module expects (up_proj, down_proj).
+        """
+        if skip_modules is None:
+            skip_modules = []
+
+        # Map HuggingFace StarCoder2 weight names to TensorRT-LLM names
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_starcoder2.py]: not 100% sure, but the parent class `DecoderModelForCausalLM` has such functionality to load weights by giving params_map, did you try that?

File: tests/unittest/_torch/modeling/test_modeling_starcoder2.py [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+                    logits = graph_runner.replay(key, inputs)
+
+            # Compare with HuggingFace
+            ref = hf_starcoder2.forward(
+                input_ids=gen_input_ids.unsqueeze(0),
+                position_ids=gen_position_ids,
+                past_key_values=ref.past_key_values,
+                use_cache=True,
+            )
+            torch.testing.assert_close(logits, ref.logits[:, -1].float(), atol=0.4, rtol=0.4)
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/modeling/test_modeling_starcoder2.py]: > If we can reasonably make these bounds tighter
Like this. 

atol=0.4 and rtol=0.4 are the default settings for other `allclose_to_hf` test so I choosed them in the first place. This test can indeed use much tighter bounds.

File: tests/unittest/_torch/modeling/test_modeling_starcoder2.py [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+        min_mem_required = {
+            "3B": 10 * (2**30),  # 10 GB
+            "7B": 20 * (2**30),  # 20 GB
+            "15B": 40 * (2**30),  # 40 GB
+        }
+
+        if total_mem < min_mem_required[config_name]:
+            self.skipTest(f"Insufficient memory for StarCoder2-{config_name}")
+
+        # Load full pretrained model from HuggingFace (same as engine test)
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/modeling/test_modeling_starcoder2.py]: Ah I was using full model weights to do a sanity check locally. Updated to use random weights.

File: tests/unittest/_torch/modeling/test_modeling_starcoder2.py [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+    _, total_mem = torch.cuda.mem_get_info()
+    # scale model down if gpu memory is low
+    if total_mem < mem_for_full_model:
+        model_fraction = total_mem / mem_for_full_model
+        num_layers = int(config_dict["num_hidden_layers"] * model_fraction)
+        num_layers = min(num_layers, default_num_layers)
+        config_dict["num_hidden_layers"] = num_layers
+
+
+class TestStarcoder2(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tests/unittest/_torch/modeling/test_modeling_starcoder2.py]: `pytest` certainly works. I was checking other model unit tests in the same directory and found that the majority of them were using `unittest`, hence my test implementation. Changed to pytest for modern and robust unittest. 

BTW I don't think it is worth wrapping the config file into a pytest.fixture. It is a simple config and we don't expect any changes after removing the `deepcopy` usecase.

File: tensorrt_llm/_torch/models/modeling_starcoder2.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+                residual=residual,
+                spec_metadata=spec_metadata,
+                lora_params=lora_params,
+            )
+
+        hidden_states = self.norm(hidden_states + residual)
+        return hidden_states
+
+
+@register_auto_model("Starcoder2ForCausalLM")
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/models/modeling_starcoder2.py]: Not sure if I understand your question, what do you mean by the default tokenizer option?

File: tensorrt_llm/_torch/models/modeling_starcoder2.py (hunk 3 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/models/modeling_starcoder2.py]: changed to TRTLLM layer_norm.

File: tensorrt_llm/_torch/models/modeling_starcoder2.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        **kwargs,
+    ):
+        if residual is None:
+            residual = hidden_states
+            hidden_states = self.input_layernorm(hidden_states)
+        else:
+            hidden_states, residual = (
+                self.input_layernorm(hidden_states + residual),
+                hidden_states + residual,
+            )
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/models/modeling_starcoder2.py]: fixed.

**Comment by yibinl-nvidia** [comment]
@2ez4bz @Wanli-Jiang all comments are addressed, could you review again? Thank you!

**Review by Wanli-Jiang — APPROVED** [review]
